### PR TITLE
Stacking Promotions

### PR DESCRIPTION
**Pull request by mamhoff**
This is a working prototype of stacking promotions. For now the only calculator that does it "right" is the distributed amounts calculator. 

The way it's done it we introduce a wrapper around the `Spree::Order`, `Spree::LineItem`, `Spree::Shipment` and `Spree::ShippingRate` objects. They get an instance variable `discounts` that holds the discounts from the previously applied lanes, and a `discounted_amount` getter that subtracts the discounts from the item's amount.

### CODE REVIEW

File: spec/models/promotion/integration_spec.rb [hunk 1]
```diff
@@ -51,6 +51,44 @@
...
+
+    let!(:distributed_amount_promo) do
+      create(:friendly_promotion,
+        :with_adjustable_action,
+        preferred_amount: 10.0,
+        apply_automatically: true,
+        lane: :post,
+        calculator_class: SolidusFriendlyPromotions::Calculators::DistributedAmount)
+    end
+    let(:promotion) { create(:friendly_promotion, name: "20% off everything", apply_automatically: true) }
```
Comment (davecandlescience) [inline spec/models/promotion/integration_spec.rb:68]: The name should be "20% off shirts" right?

File: spec/models/solidus_friendly_promotions/promotion_action_spec.rb [hunk 2]
```diff
@@ -18,27 +18,28 @@
...
     let(:promotion) { SolidusFriendlyPromotions::Promotion.new(name: "20 Perzent off") }
     let(:action) { described_class.new(promotion: promotion) }
 
     before do
       allow(action).to receive(:compute_amount).and_return(-1)
     end
 
-    it "returs an discount to the adjustable" do
-      expect(subject).to eq(
+    it "returs an discount to the discountable" do
```
Comment (davecandlescience) [inline spec/models/solidus_friendly_promotions/promotion_action_spec.rb:34]: returns

File: app/models/solidus_friendly_promotions/friendly_promotion_discounter.rb [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 module SolidusFriendlyPromotions
   class FriendlyPromotionDiscounter
-    attr_reader :order, :promotions, :item_discounter
+    attr_reader :order, :promotions
```
Comment (davecandlescience) [inline app/models/solidus_friendly_promotions/friendly_promotion_discounter.rb:5]: Remove the @item_discounter on line 10?

File: spec/models/promotion/integration_spec.rb (hunk 1 above)
Comment (mamhoff) [reply spec/models/promotion/integration_spec.rb]: Fixed

**Review by davecandlescience — APPROVED** [review]
LGTM

**Pull request merged**